wiki/src/main/resources/XWiki/PDFViewerMacro.xml

```diff
@@ -47,4 +47,138 @@
+        <freeText>forbidden</freeText>
+        <largeStorage>0</largeStorage>
+        <multiSelect>0</multiSelect>
+        <name>cache</name>
+        <number>5</number>
+        <prettyName>Caching policy</prettyName>
+        <relationalStorage>0</relationalStorage>
+        <separator> </separator>
+        <separators>|, </separators>
+        <size>1</size>
+        <unmodifiable>0</unmodifiable>
+        <values>long|short|default|forbid</values>
+        <classType>com.xpn.xwiki.objects.classes.StaticListClass</classType>
+      </cache>
+      <code>
+        <contenttype>PureText</contenttype>
+        <disabled>0</disabled>
+        <editor>PureText</editor>
+        <name>code</name>
+        <number>2</number>
+        <prettyName>Code</prettyName>
+        <rows>20</rows>
+        <size>50</size>
+        <unmodifiable>0</unmodifiable>
+        <classType>com.xpn.xwiki.objects.classes.TextAreaClass</classType>
+      </code>
+      <name>
+        <disabled>0</disabled>
+        <name>name</name>
+        <number>1</number>
+        <prettyName>Name</prettyName>
+        <size>30</size>
+        <unmodifiable>0</unmodifiable>
+        <classType>com.xpn.xwiki.objects.classes.StringClass</classType>
+      </name>
+      <parse>
+        <disabled>0</disabled>
+        <displayFormType>select</displayFormType>
+        <displayType>yesno</displayType>
+        <name>parse</name>
+        <number>4</number>
+        <prettyName>Parse content</prettyName>
+        <unmodifiable>0</unmodifiable>
+        <classType>com.xpn.xwiki.objects.classes.BooleanClass</classType>
+      </parse>
+      <use>
+        <cache>0</cache>
+        <disabled>0</disabled>
+        <displayType>select</displayType>
+        <freeText>forbidden</freeText>
+        <largeStorage>0</largeStorage>
+        <multiSelect>0</multiSelect>
+        <name>use</name>
+        <number>3</number>
+        <prettyName>Use this extension</prettyName>
+        <relationalStorage>0</relationalStorage>
+        <separator> </separator>
+        <separators>|, </separators>
+        <size>1</size>
+        <unmodifiable>0</unmodifiable>
+        <values>currentPage|onDemand|always</values>
+        <classType>com.xpn.xwiki.objects.classes.StaticListClass</classType>
+      </use>
+    </class>
+    <property>
+      <cache>long</cache>
+    </property>
+    <property>
+      <code>require(['jquery'], function ($) {
+  $(function() {
+    let viewerFrame = $('iframe.pdfviewer');
+    viewerFrame.on('load', function() {
+      let secondaryMenuButton = viewerFrame.contents().find('#secondaryToolbarToggle');
+      // Triggered when clicking the copy button.
+      let copyPdf = function () {
+        let fileName = new URL(viewerFrame.attr('src'), location.origin).searchParams.get('file');
+        if (!fileName.includes('http://') &amp;&amp; !fileName.includes('https://')) {
+          fileName = location.origin + fileName;
+        }
```

Comment:
You resolve above the iframe src relative to the window location. Why not do the same with the PDF file name?

```suggestion
        let pdfURL = new URL(fileName, location.origin).toString();
```

Reply:
The PDF file name can also be an external link.

Reply:
Yes, in which case it means the file name is an absolute URL and ``new URL(absoluteURL, anotherAbsoluteURL)`` returns ``absoluteURL``. See https://developer.mozilla.org/en-US/docs/Web/API/URL/URL :

> base: A USVString representing the base URL **to use in cases where url is a relative URL**.

If the first argument is an absolute URL then the second argument is simply ignored.